### PR TITLE
[test] add tests for DNS forwarding

### PR DESCRIPTION
**Pull request by erjiaqing**
This commit adds a new file `tests/scripts/thread-cert/border_router/nat64/test_upstream_dns.py` to test scripts for testing the DNS forwarding functions.

The test adds a record `test.domain` to the bind9 interface since when adding to DNSSD tests, the OpenThread's DNSSD server will be conflict with the Docker's DNS forwarding functions.

### CODE REVIEW

File: tests/scripts/thread-cert/border_router/nat64/test_upstream_dns.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+BR = 1
+ROUTER = 2
+HOST = 3
+
+TEST_DOMAIN = 'test.domain'
+TEST_DOMAIN_IP6_ADDRESSES = {'2001:db8::1'}
+
+
+class Nat64SingleBorderRouter(thread_cert.TestCase):
```
Comment (superwhd) [inline tests/scripts/thread-cert/border_router/nat64/test_upstream_dns.py:56]: Should we change the name to something like `Nat64UpstreamDns`?

File: tests/scripts/thread-cert/border_router/nat64/test_single_border_router.py [hunk 2]
```diff
@@ -63,6 +63,9 @@
 # So the BR will remove the random-generated one.
 LARGE_NAT64_PREFIX = "ff00:00:00:01:00:00::/96"
 
+GOOGLE_DNS_DOMAIN = 'dns.google'
+GOOGLE_DNS_IP6_ADDRESSES = {'2001:4860:4860::8888', '2001:4860:4860::8844'}
```
[the remark this answers is not in the document]
Comment (erjiaqing) [reply tests/scripts/thread-cert/border_router/nat64/test_single_border_router.py]: Forgot to remove from draft. Removed now.

File: src/posix/platform/resolver.cpp [hunk 3]
```diff
@@ -43,6 +43,7 @@
 #include <netinet/in.h>
 #include <sys/select.h>
 #include <sys/socket.h>
+#include <unistd.h>
```
[the remark this answers is not in the document]
Comment (erjiaqing) [reply src/posix/platform/resolver.cpp:46]: Seems the build may fail for missing functions (missing `read`) on different toolchains without this.

File: tests/scripts/thread-cert/border_router/nat64/test_upstream_dns.py (hunk 1 above)
Comment (erjiaqing) [reply tests/scripts/thread-cert/border_router/nat64/test_upstream_dns.py]: Changed to `UpstreamDns`

**Review by superwhd — APPROVED** [review]
LGTM 👍

File: tests/scripts/thread-cert/border_router/nat64/test_upstream_dns.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        br.bash(
+            shlex.join(['echo', f'zone "{TEST_DOMAIN}" {{ type master; file "/etc/bind/db.test.domain"; }};']) +
+            ' >> /etc/bind/named.conf.local')
+        br.bash(shlex.join(['echo', '$TTL 24h']) + ' >> /etc/bind/db.test.domain')
+        br.bash(
+            shlex.join(['echo', f'@ IN SOA {TEST_DOMAIN} test.{TEST_DOMAIN}. ( 20230317 86400 300 604800 3600 )']) +
+            ' >> /etc/bind/db.test.domain')
+        br.bash(shlex.join(['echo', f'@ IN NS {TEST_DOMAIN}.']) + ' >> /etc/bind/db.test.domain')
+        for addr in TEST_DOMAIN_IP6_ADDRESSES:
+            br.bash(shlex.join(['echo', f'@ IN AAAA {addr};']) + ' >> /etc/bind/db.test.domain')
```
Comment (xuyirio) [inline tests/scripts/thread-cert/border_router/nat64/test_upstream_dns.py:106]: Could we wrap these into a function for better readability?

**Review by xuyirio — APPROVED** [review]
Thanks, Song!

File: tests/scripts/thread-cert/border_router/nat64/test_upstream_dns.py (hunk 4 above)
Comment (erjiaqing) [reply tests/scripts/thread-cert/border_router/nat64/test_upstream_dns.py]: extracted this into a single string block, hope it has better readability now

**Review by jwhui — APPROVED** [review]

**Pull request merged**